Blocks: Update default block categories

aduth opened the pull request:
Closes #11406

This pull request seeks to propose to replace the existing set of default block categories with a new set of categories as proposed in #11406.

The updated set of block categories are:

- Text
- Media
- Design
- ~Tools~ **Edit:** Widgets _(Unchanged)_
- Embeds _(Unchanged)_
- Reusable _(Unchanged)_

![image](https://user-images.githubusercontent.com/1779930/71267560-e23c2080-2318-11ea-8eee-c56e8eb26db4.png)

<p><strong id="implementation-notes">Implementation Notes:</strong></p>

There is an attempt to preserve some backwards-compatibility here, based on the summarized findings at https://github.com/WordPress/gutenberg/issues/11406#issuecomment-567164861 .

Blocks which register to one of the existing categories will transparently map to one of the updated categories. As implemented, that mapping is proposed as follows:

Before|After
---|---
Common|**_Text_**
Formatting|**_Text_**
&nbsp;|Media
Layout|**_Design_**
Widgets|Widgets
Embeds|Embeds
Reusable blocks|Reusable blocks

<details><summary>Click to expand outdated notes from early iterations</summary>

Before|After
---|---
Common|Text
Formatting|Text
Layout|Design
Widgets|Tools

_(Suggested improvements to this are welcomed)_

Notably, I have not yet updated existing blocks to use an appropriate new category name, for the explicit purpose to effectively test the sensibility of this mapping arrangement, considering that this is expected to impact third-party blocks which register to one of the existing categories, at least until those plugins opt to update to a more appropriate updated category.

Prior to merge, we can consider to update the default set of core blocks to a more appropriate block category.
</details>

~As a follow-up task, I would like to explore improvements around allowing blocks to be registered without a category, or without a valid category. Such blocks should be shown in an "Uncategorized" block grouping. This could also be used to support usage where there are no default categories, and the inserter would show all blocks without any groupings. This could also be used to improve support for hypothetical future revisions to block categories, in order to better handle block registrations for plugins which support multiple older versions of WordPress.~ **Edit (2020-05-29):** This was implemented separately in #22280.

**Testing Instructions:**

Verify block categories are updated to new values, seen in the block inserter.

Verify there are no errors of block registration in your browsers Developer Tools console (i.e. all blocks should still be registered, with an appropriate category).

Optionally, verify you can change a block registration to use one of the new block category names, or a legacy category name. Legacy category usage should appear as categorized based on the mapping above.

aduth committed:
Blocks: Update default block categories

aduth committed:
Block API: Add wp.blocks.getCategory singular function

ZebulanStanphill commented:
I think "Tools" is an unclear category name. I have no idea what kind of block would be a "tool". I think it should be replaced with an "Uncategorized" category for blocks that don't fit anywhere else.

youknowriad commented on packages/blocks/src/api/categories.js:
```diff
@@ -3,19 +3,32 @@
+ *
+ * @return {WPBlockCategory|undefined} Block category, if exists.
+ */
+export function getCategory( slug ) {
```
Can we avoid the new global functions, I'm thinking we should favor selectors and registry selectors as much as we can to avoid the "singleton" issue. We definitely have this issue in a lot of existing functions but having less global functions will force us to use selectors for these more.

aduth replied:
> Can we avoid the new global functions, I'm thinking we should favor selectors and registry selectors as much as we can to avoid the "singleton" issue. We definitely have this issue in a lot of existing functions but having less global functions will force us to use selectors for these more.

I agree with the general idea. For me, it was added purely for the sake of "consistency for consistency's sake", in that intentionally introducing inconsistency for when these global functions do and don't exist could likely lead to some confusion and uncertainty.

I don't feel particularly strongly about it, and can remove it if it's undesirable.

youknowriad replied:
I don't mind it being added for now but how can we discourage its usage in our code base.

aduth replied:
> I don't mind it being added for now but how can we discourage its usage in our code base.

Something like this, perhaps?

https://astexplorer.net/#/gist/1600f6a5862962ed4708045a7742dfeb/1be4a55a8de2a91b790352d1a1c01161c8d411a4

https://github.com/WordPress/gutenberg/blob/8c307ba/.eslintrc.js#L57

aduth replied:
In comparing the alternative approaches between #22192 and #22280, where the latter avoided the addition of a new selector, I wonder if it's something we can do here as part of the registration. Namely: Avoid `getCategory`, and let `registerBlockType` handle the category normalization.

aduth replied:
> In comparing the alternative approaches between #22192 and #22280, where the latter avoided the addition of a new selector, I wonder if it's something we can do here as part of the registration. Namely: Avoid `getCategory`, and let `registerBlockType` handle the category normalization.

Implemented in f1f59ee.

mapk commented:
As we're re-labeling these categories, and making minor modifications (global functions), it's also important to keep in mind that we're [looking to add Block Patterns to the Inserter](https://github.com/WordPress/gutenberg/issues/17335) too. This will likely include a new set of categories that apply to patterns. Reusing this code for that makes a lot of sense.

shaunandrews commented:
Works well in my testing. 👍

mapk commented:
In relation to https://github.com/WordPress/gutenberg/issues/19409, it's apparent that "Tools" may not work here. It's seem to be our goto word lately and is being used a bit too much.

**In the Tools/Modes popover**

![Screen Shot 2020-01-16 at 8 30 38 AM](https://user-images.githubusercontent.com/617986/72543634-e46f9d00-383a-11ea-9ad3-5fb611dea695.png)

**In the Options popover**

![Screen Shot 2020-01-16 at 8 30 46 AM](https://user-images.githubusercontent.com/617986/72543654-eafe1480-383a-11ea-8e50-3344cce6128c.png)

**And also here in the Inserter** (if we make this change)

![Screen Shot 2020-01-16 at 8 32 24 AM](https://user-images.githubusercontent.com/617986/72543685-f5201300-383a-11ea-9290-62c851c5a27f.png)


It looks like "Tools" includes
`Latest Posts, Search, Read More, Calendar, Archives, Categories, Shortcode, Page Break, Classic.`

There was a suggestion from @mtias to possibly call it "Blog" but "Tools" had felt better at the time. Now I'm rethinking Blog.

ZebulanStanphill commented:
Why not just call it "Other" or "Uncategorized"? There's going to be a need for an "other"-type category anyway, so why not just use that instead of a contrived but still ambiguous name like "Tools"?

karmatosed commented:
I agree with not using 'tools' as that is becoming a repeating word that I suspect will confuse.  I'd lean away from creating buckets before they are needed so 'uncategorized' or 'other' to me wouldn't be ideal. Blog seems to work as does 'element' maybe?

shaunandrews commented:
> Latest Posts, Search, Read More, Calendar, Archives, Categories, Shortcode, Page Break, Classic.

Everything (but Classic) makes sense under "Blog" to me.

Maybe the Classic block fits better under Text?

ZebulanStanphill commented:
Neither Search nor Shortcode are really related to the concept of "blog" in my opinion. Classic belongs wherever Custom HTML belongs, in my opinion.

pyronaur commented:
> Everything (but Classic) makes sense under "Blog" to me.

It makes sense to me too. However, does it make sense to someone who's not familiar with blogging/WordPress - is Calendar really a part from a blog? For example, if their day-to-day vocabulary is mobile and desktop software - Calendar is an app, not a part of "Blog Blocks" or even Components.

>  I'd lean away from creating buckets before they are needed so 'uncategorized' or 'other' to me wouldn't be ideal

I think this might already be a good time for a creating that bucket. Especially considering custom blocks.

What if someone is creating a plugin with a single block that does something that's not a component or interface? If they find no "Other" category where to place their block in - they might resort to creating the other category themselves anyway, and at that point - it might be even more confusing for what really is "Blog" or "Component" and what is "Other"/"Misc".

Also, considering naming - "Other/Misc/Miscellaneous/Not a Block/Uncategorized" - if there's no standardized bucket for this - someone might end up with a blog where they have both "Other" and "Misc" categories because different developers decided to use different naming conventions.

boemedia commented:
If we’re discussing ‘blog’, i’d avoid it in this context. It will be confusing if blocks are used on other types of websites or in general in pages.

MichaelArestad commented:
Howdy! It looks like we're stuck on renaming Widgets.

It seems to me we could do one of a few things:
1. Make a few more categories instead of trying to make blocks fit where they don't. This feels like the most logical and scalable solution to me. It also encourages meaningful categorization. Here's a pass at this:
```
Text
Media
Layout (assuming we don't call Templates Layouts)
Post - Latest posts, Latest comments
Sharing - Sharing buttons, RSS
Navigation - Navigation, Search, Archives, Categories, Tag Cloud
Embeds
Reusable
```
2. Use Blog - While it fits okay, it seems to me if I wasn't making a blog, I'd never even open that category. I don't want those poor little blocks to feel left out.
3. Use Widgets, Tools, Elements, or Uncategorized - These are all identical to me. They represent a junk drawer. I don't recommend any of them. Instead, let's encourage developers to put their blocks in a category, even if they have to make a special category just for their cool block. If we create a miscellaneous category, it will fill up pretty quickly.
4. Leave it as Widgets for now while debating the verbiage in a follow-up PR.

Phew! Having said all that, these need further exploration. @melchoyce mentioned that a card sorting exercise might be valuable here. Would anyone care to take a shot at it?

karmatosed commented:
This is all very tricky to find a grouping that works across so deeply appreciate everyone working together here to get there. I do find the 'Navigation' grouping a little surprising as I don't know if I feel that categories and tag cloud to me are navigation, I wonder if it could be 'elements'?

pyronaur commented:
@karmatosed What are the downsides to creating the "Other" bucket? Just curious about the hesitance there.

boemedia commented:
> @karmatosed What are the downsides to creating the "Other" bucket? Just curious about the hesitance there.

It can easily become a trash can of too many blocks.

boemedia commented:
> This is all very tricky to find a grouping that works across so deeply appreciate everyone working together here to get there. I do find the 'Navigation' grouping a little surprising as I don't know if I feel that categories and tag cloud to me are navigation, I wonder if it could be 'elements'?

Aren't all blocks 'elements' in a way?

boemedia commented:
Would it be an idea to step away from category names for now and look at the blocks from a purpose perspective? Maybe this approach helps categorising them better and, eventually, come up with a good name per block category. 

So: The ______ block is for _______

or

With the ______ block you can ________


Is there a document with a list of all current native blocks?

pyronaur commented:
> It can easily become a trash can of too many blocks.

Totally can see that happening. But perhaps they can then be gradually categorized out of there as the need crops up naturally?

It seems to me that, currently, we're trying to figure out a word for "Other" that doesn't say "Other" anyway.

pyronaur commented:
> Is there a document with a list of all current native blocks?

I'm not sure if there's a document with a list, but you can view the list as directories here:
https://github.com/WordPress/gutenberg/tree/master/packages/block-library/src

MichaelArestad commented:
> This is all very tricky to find a grouping that works across so deeply appreciate everyone working together here to get there. I do find the 'Navigation' grouping a little surprising as I don't know if I feel that categories and tag cloud to me are navigation, I wonder if it could be 'elements'?

@karmatosed Are you suggesting "Navigation elements" or just "elements"?

MichaelArestad commented:
> It seems to me that, currently, we're trying to figure out a word for "Other" that doesn't say "Other" anyway.

I'm leaning against that. Even if we have a category with one or two blocks, it sets a good example and other block creators can add their blocks to that category. For example, in my earlier comment, I proposed a potential "Sharing" category. That's one that is easy to see how a variety of other blocks could be added into that category as sharing widgets/plugins are pervasive and valuable.

ZebulanStanphill commented:
I agree that "tools", "widgets", and "elements" are all just unnecessarily fancy ways of saying "other".

Regarding the idea of "post" and "navigation" categories, it seems a little weird to have the Latest Posts and Archives blocks be in separate categories, because they're really quite similar. They're both a list of links. And I'd argue that Latest Posts is just as much a navigation block as Archives. The same goes for the Latest Comments block.

Now here's a tough one: where do you put the Legacy Widget block? It can contain all sorts of widgets that might be text-related, media-related, or both. It's kind of like the Classic block in that it exists for backwards compatibility. But does it really make sense to create a whole category for blocks that exist solely for compatibility reasons? I don't think so. And besides, the Classic block feels more like a Text block to me.

I wonder if maybe the whole category system needs to be reconsidered? Maybe we should be using a tag-based system instead where blocks can be given multiple different tags that you can filter search results by, rather than forcing blocks to only belong to one category.

If you're going to avoid an "other" category, you have to figure out where weird blocks like these belong:

- Classic
- Custom HTML
- Legacy Widget
- More
- Page Break
- RSS
- Shortcode

pyronaur commented:
> I'm leaning against that. Even if we have a category with one or two blocks, it sets a good example and other block creators can add their blocks to that category.

@MichaelArestad  - If there's no Other category -  is there a good way we could we address my concern above? 

> Also, considering naming - "Other/Misc/Miscellaneous/Not a Block/Uncategorized" - if there's no standardized bucket for this - someone might end up with a blog where they have both "Other" and "Misc" categories because different developers decided to use different naming conventions.

karmatosed commented:
> Are you suggesting "Navigation elements" or just "elements"?

Just elements.

karmatosed commented:
> What are the downsides to creating the "Other" bucket? Just curious about the hesitance there.

As @boemedia noted, it becomes a catch-all space where things fall. I would like us if we use categories to use them over have to catch ones that don't fit.

ZebulanStanphill commented:
@karmatosed
> > Are you suggesting "Navigation elements" or just "elements"?
> 
> Just elements.

How would you define "elements"? The term is already used for HTML elements, so it seems rather confusing to me.

mtias commented:
I've been going back at this as I feel categories are causing hurdles rather than helping and had some similar thoughts to @boemedia. Mine were more about thinking in verb form as a way to organize intent. The other aspect is that we are going to be getting a surge in blocks with full-site editing that has a patently different intent. So I propose a reduction in categories to the following:

- **Create** (Text, Media, and anything related to composing).
- **Build** (widgets, theme blocks for FSE, layout, etc).
- **Share** (embeds, social icons, some widgets like RSS).

No matter what taxonomy we come up with, search will be a primary way to discover and find blocks, so I'd suggest keeping categories short and expressive while we continue to optimize search.

Plugins would already be able to leverage _collections_ to group based on source (needs to be presented differently in the inserter, but that is a design task).

Finally, I think each of these groups should have the frequent + recent mechanism to sort things, so that if you always write with Classic or mainly publish images, that'd be at the top of Create.

mapk commented:
> So I propose a reduction in categories to the following:

Yep, this helps deemphasize the category names and eases the current difficulty of jumping through various accordions to find the right block.

Changing the category names to verbs works for me too. While these don't initially explain which blocks the user may find, they're generalized and aimed toward actions. This is good. However, when plugins add their categories to the Library, they most likely won't follow the "verb" standard and just list their plugin name. I don't think that should be a concern though.

Here's how I see the blocks could be categorized based on @mtias' suggestion.


Category | Blocks
------------ | -------------
Most&nbsp;Used | (Most used blocks)
Create | Paragraph, Heading, List, Quote, Pull Quote, Table, Verse, HTML, Code, Preformatted, Image, Gallery, Audio, Video, File
Build | Columns, Group, Cover, Media+Text, Buttons, Separator, Spacer, Navigation, Latest Posts, Latest Comments, Search, More, Calendar, Archives, Categories, Shortcode, Page Break, Tag Cloud, Classic, _Theme blocks for FSE_, _Post blocks_
Share | Social Icons, RSS, _Embeds_

![Block Library](https://user-images.githubusercontent.com/617986/74849890-22eef080-52ee-11ea-981a-ece346f07108.png)

mtias commented:
To clarify, this proposal would drop "Most Used" as well, as each group would be organized taking usage into account.

> However, when plugins add their categories to the Library, they most likely won't follow the "verb" standard and just list their plugin name. I don't think that should be a concern though.

This should be part of education. Plugins don't need to register a category with their name because they can use _collections_, which are tied to the namespace of the block.

ZebulanStanphill commented:
"Create" and "Build" really just sound like the same thing to me, and the lines between where something should go really don't feel that clear even for "Share". A YouTube embed really isn't that different from a Video block, so why are they in different categories? You could argue a `<video>` or `<img>` element is still "embedding" something into the page. You can even use external URLs in some of the standard media blocks.

Also, I don't think each category should be sorted by usage. I've found that sorting by usage tends to make finding blocks more confusing since it prevents muscle memory from forming by constantly reorganizing blocks. I prefer alphabetical order.

mtias commented:
That's a fair point, but I don't think it matters too much in practice. These categories will never be entirely adequate unless we increase their number significantly — or create generic buckets like "Others" — to avoid all misplacements, and at that point it becomes a tedious layer to sort through to browse blocks. A flat list would almost be better.

The appeal of just Create, Build, and Share is that it gives a rough split between editing and site building, which is a useful mental model in our context. It doesn't mean you can't use pieces of build when composing a post, but it gives us a better place to group a lot of the incoming new blocks.

Embeds can be seen as content, for sure, and that's why things like "video" let's you drop a youtube url. But I believe it's still useful to separate them, cognitively, as there are already too many embeds in place and they deal with third-party services in some capacity.

Finally, less clutter in categories means we can focus more on optimizing the search interface, which is going to be the primary way to find and sort through the growing number of blocks.

I'm less sold on sorting by usage because I agree it tends to annoy more than help people.

pyronaur commented:
I like the new approach a lot better. But I also have to agree with @ZebulanStanphill - what makes a something be "Create" or "Build"? Perhaps that's something that can be clarified in the category name? 

For example, if guessing the category based on language as spoken, I'd find it difficult to figure out these:
* build column vs create column
* create table vs build table

I had to look this up:

> define create
> bring (something) into existence.
> 
> define build
> construct (something) by putting parts or material together.

Create & Build make sense now that I know the difference in definition. For a first time WordPress user - that might not be the case.

Also, after translating those words to my native language becomes even more difficult to figure out what the difference is (perhaps @shinyabw has insight on this too? ).

Naming things is hard 🤪

---

What about something like this?

| Category | Blocks |
| -------------------- | ------------- |
|Rich Text | Paragraph, Heading, List, Quote, Pull Quote, Table, Verse, Code, Preformatted, Image, Gallery, Audio, Video, File|
|Layout | Columns, Cover, Media + Text, Buttons, Separator, Spacer, Page Break|
|Dynamic | HTML, Navigation, Latest Posts, Latest Comments, Search, More, Calendar, Archives, Categories, Shortcode, Tag Cloud, RSS, Embeds, Social Icons|


No idea where to place: Classic, Theme blocks for FSE, Post blocks

Or maybe add an additional category: "Elements" ?

|  Category | Blocks |
| ----------------------- | ------------- |
|Rich Text | Paragraph, Heading, List, Quote, Pull Quote, Table, Verse, Code, Preformatted, Image, Gallery, Audio, Video, File|
|Elements | Buttons, Embeds, Social Icons|
|Layout | Columns, Group, Cover, Separator, Spacer, Page Break, Media + Text|
|Dynamic | HTML, Navigation, Latest Posts, Latest Comments, Search, More, Calendar, Archives, Categories, Shortcode, Tag Cloud, RSS|


**Rich Text** - The default elements that the user is likely expecting to be a part of any Rich Text Editor (Word, TinyMCE, etc.)
**Elements** - Web Specific elements (sharing, embeds, etc.)
**Layout** - Columns, Groups, Lines, etc.
**Dynamic** - Blocks with content that isn't written by hand

boemedia commented:
I agree with @pyronaur that the difference between 'build' and 'create' is hard to get. And will indeed give problems when translating into other languages.

I'd be careful with using categories like 'dynamic' or 'rich text'. This doesn't mean anything to non-technical users. "Elements' is so generic, it's really hard to guess what's in that category.

Instead of keeping this discussing in this ticket, I'd advocate for user testing eg. with card sorting and see what users come up with. This could be done at Meetups, WordCamps... ideally with participants from all kinds of user groups, but at least with non-tech users.

boemedia commented:
Quickly looked at how Wix handles page/blog elements. Their menu is more extensive and als separated into more specific categories. See the video I created with Cloudapp [https://cl.ly/d81e8e46351e](https://cl.ly/d81e8e46351e).

In comparison, I also looked at how Squarespace has categorised the editor. This one is a lot less intuitive and als contains a category 'more', that I think should be avoided as discussed before. https://cl.ly/1104ef1bc693

mtias commented:
@boemedia the Wix version looks like it optimizes insertion for what we are calling patterns. I agree that for those it makes more sense to have (almost) a category per block type (text, images, galleries, lists, buttons, etc). Exploring something like that for the patterns library in #17335 would be good.

We need to separately simplify the block categories because it's getting out of hand:

![image](https://user-images.githubusercontent.com/548849/76168625-cc006c80-6171-11ea-92b7-67af822fea13.png)

Side note, but the collapsing / expanding panels hide a lot of context by not showing the blocks by default, which makes the categories even more daunting to browse.

mtias commented:
I've reviewed the conversations again with some distance and aiming to summarize next steps. The last proposal (create, build, share) seemed too abstract for people, so going back a step to the conclusion from #11406 — let's do 5 categories: Text, Media, Design, Tools/Widgets, Embeds.

I'd probably lean on "widgets" since it has more acumen in the WordPress world. For FSE we could consider more categories to organize all the template tags that are surfacing as blocks.

Embeds remains as a separate category because it's a grouping that can grow and scale significantly but it should not get in the way of other blocks. I would suggest lifting blocks that are seeing high use out of that category and into the media one (like YouTube video).

![image](https://user-images.githubusercontent.com/548849/80620366-3208c200-8a46-11ea-82b9-347583c7c667.png)

mtias commented:
Oh, one more thing, I think the "Most Used" category should be kept internally (not in the inserter) and use it to power the priority queue of the `/` inserter.

ZebulanStanphill commented:
I definitely like (Text, Media, Design, Widgets, and Embeds) more than the current category divisions and the previous super-abstract proposal. And I agree with taking the Most Used category out of the standard inserter.

I think we should probably separate collections from categories. Perhaps a dropdown filter to only show blocks from the selected collection?

mtias commented:
Yes, collections should definitely be distinguished from block categories. That's what will allow plugins to register blocks in the most semantic place while keeping a view for all the blocks from the plugin if they want:

<img width="516" alt="77783236-dd30f080-702e-11ea-9255-3c7d3c7df39f" src="https://user-images.githubusercontent.com/548849/80628308-72217200-8a51-11ea-8a29-c5097760dfb8.png">

mapk commented:
@aduth I just tested this on gutenberg.run and it looks like the category `Media` is missing and `Most used` is still in play.

Currently, I see:

- Most used
- Text
- Design
- Tools 
- Embeds

**But it should be:**

- **Text**
- **Media**
- **Design**
- **Widgets**
- **Embeds**

(I still don't like `widgets` and `tools` is overused in our UI, but we can't seem to settle on anything else)

aduth commented:
I've been working to bring this back up-to-date today, using the [latest suggestions](https://github.com/WordPress/gutenberg/pull/19279#issuecomment-621322775) as a base. There's a lingering end-to-end test failure which appears to be based on what I feel to be an inconsistent behavior of the block autocompleter. I might plan to address this as a separate blocking effort, since this pull request is already quite large and unwieldy.

In any case, I'd welcome some feedback on how it feels currently.

Notably, here's a consolidated list of the current "mappings" of old to new category:

Before|After
---|---
Common|**_Text_**
Formatting|**_Text_**
&nbsp;|Media
Layout|**_Design_**
Widgets|Widgets
Embeds|Embeds
Reusable blocks|Reusable blocks

In practice, this means the following classifications of blocks:

**Text**

- Paragraph
- Image
- Heading
- Gallery
- List
- Quote
- Audio
- Code
- Column
- Cover
- File
- Classic
- Custom HTML
- Preformatted
- Pullquote
- Table
- Verse
- Video

**Widgets**

- Shortcode
- Archives
- Calendar
- Categories
- Latest Comments
- Latest Posts
- RSS
- Search
- Social Icons
- Social Icon
- Tag Cloud

**Design**

- Button
- Buttons
- Columns
- Group
- Media & Text
- More
- Page Break
- Separator
- Spacer
- Navigation
- Navigation Link

**Embed**

- Embed
- Twitter
- YouTube
- Facebook
- Instagram
- WordPress
- SoundCloud
- Spotify
- Flickr
- Vimeo
- Animoto
- Cloudup
- Crowdsignal
- Dailymotion
- Hulu
- Imgur
- Issuu
- Kickstarter
- Meetup.com
- Mixcloud
- Reddit
- ReverbNation
- Screencast
- Scribd
- Slideshare
- SmugMug
- Speaker Deck
- TikTok
- TED
- Tumblr
- VideoPress
- WordPress.tv
- Amazon Kindle

<!-- lodash.map( lodash.groupBy( wp.blocks.getBlockTypes(), 'category' ), ( blocks, category ) => `**${ lodash.upperFirst( category ) }**\n\n${ blocks.filter( ( block ) => lodash.get( block.supports, [ 'inserter' ], true ) ).map( ( block ) => '- ' + block.title ).join( '\n' ) }` ).join( '\n\n' ) -->

Two notes:

- Regarding "Most used": This pull request is massive and I'd like to address its removal separately in a follow-up pull request.
- Regarding the apparent absence of "Media": As based on the mapping above, there are no blocks which fall into this category at the moment, hence why it doesn't appear.

Reiterating an early point, the mapping exists as a means to ensure compatibility for existing blocks, to give _something_ as a fallback for plugins which haven't updated to use this new categorization. We can still rearrange core blocks to more appropriate categories. I was intentional in omitting this in initial iterations pending further feedback, since I think it serves as an example of how we can expect this to play out for existing plugin blocks.

At this point, I could use feedback about any specific blocks which should be recategorized. To the previous point, note that each action we take in recategorizing a block effectively amounts to a third-party plugin developer needing to do the same.

Two further thoughts:

- We don't really have a good "story" yet for how a plugin who supports older versions of WordPress or Gutenberg should plan to update their categories, since the new categories will of course not exist in those older versions.
- One option for an alternative to a "mapping" is that we could allow the existing categories to exist, but add the new ones, transition all core blocks, and hope that plugin authors will transition as well, to a point where the old category effectively disappears by the absence of any block assigned to that category.

mapk commented:
@aduth thanks for the update!

> Regarding "Most used": This pull request is massive and I'd like to address its removal separately in a follow-up pull request.

Sounds good.

> Regarding the apparent absence of "Media": As based on the mapping above, there are no blocks which fall into this category at the moment, hence why it doesn't appear.

We'll need to remap some to the `Media` category.

>  I could use feedback about any specific blocks which should be recategorized.

Move from `Text` to `Media`

- Image
- Gallery
- Audio
- Cover
- File
- Video

- Remove Column block (this shouldn't be visible outside the Columns block)

Move from `Text` to `Widgets`

- Classic

Move from `Design` to `Media`

- Media+Text

aduth commented:
>Remove Column block (this shouldn't be visible outside the Columns block)

Ah, I don't think it's actually there. I generated the lists above using an ad hoc script, and I neglected to account for blocks which are restricted by parent block (Column only inside a Columns).

ZebulanStanphill commented:
Here are the blocks I think should be moved:

- **Audio**: Text → Media
- **Buttons**: Design → Widgets
- **Classic**: Text → Widgets
- **Cover**: Text → Design
- **Custom HTML**: Text → Widgets
- **File**: Text → Media
- **Gallery**: Text → Media
- **Image**: Text → Media
- **Navigation**: Design → Widgets
- **Video**: Text → Media

mtias commented:
This is great, thanks @aduth. Agreed with moving those to **Media**. @mapk I'm also not super happy with **Widgets** (or its alternative, **Tools**), but since it's a common term in WP parlance, and where we'll actually show most of the blocks _that used to be widgets_, I think we can settle on it for now otherwise we'll continue debating this one endlessly :)

@youknowriad has mentioned a few times it might be good to have a fallback to "uncategorized" to avoid blocks not showing up if a category is unregistered or not present (by any mechanism). I think it might make sense to introduce that behaviour here or soon after so that it goes in the same release. It'll help make future changes knowing there will always be a safe net.

Since we are keeping "Most Used" to power auto-completion, a separate task might also be to consider allowing users to toggle that category on / off in "Settings". (And I think we should reduce it to 6 as well as a maximum; more than that and it doesn't feel like most used anymore.)

aduth commented:
> @youknowriad has mentioned a few times it might be good to have a fallback to "uncategorized" to avoid blocks not showing up if a category is unregistered or not present (by any mechanism). I think it might make sense to introduce that behaviour here or soon after so that it goes in the same release. It'll help make future changes knowing there will always be a safe net.

Do you see this as being the substitute for the mapping behavior currently implemented here? Or more of a solution to the problem of as I described in https://github.com/WordPress/gutenberg/pull/19279#issuecomment-624911452 of "how a plugin who supports older versions of WordPress or Gutenberg should plan to update their categories, since the new categories will of course not exist in those older versions"?

I expect that it would require a change to make `category` an optional field of block settings (it's currently required). This is something I've wanted to change for a while now. I'm not too sure what sorts of cascading effects there might be. It might be the sort of thing to do as a separate "blocking" pull request to support this one.

mtias commented:
> Do you see this as being the substitute for the mapping behavior currently implemented here?

No, I think the mapping is fine and needed, that'd be more complementary. But it'd be nice to do it in the same release.

mtias commented:
I also agree with the suggestions above, except I think we should retain "Buttons" in design, "Classic" in text (for now), and "Navigation" in design (until we figure out the best place for FSE blocks).

ZebulanStanphill commented:
> I also agree with the suggestions above, except I think we should retain "Buttons" in design, "Classic" in text (for now), and "Navigation" in design (until we figure out the best place for FSE blocks).

My thinking is that:

- Design: stuff that involves the general structure/layout of the content. Things like page breaks, columns, containers, spacers, etc. (Cover is essentially an advanced group, so I think it belongs in the same category.)
- Widgets: anything that doesn't fall into the other categories. I guess I'm treating this category as the equivalent of an "Other" category. Actually, maybe we should just rename it to that.

If Archives belongs in Widgets, why not also Navigation? They're both a list of links.

Similarly, if Social Icons belongs in Widgets, then Buttons (almost the same thing) belong there as well.

I can understand keeping Classic in Text for now for the sake of making it more visible to users more familiar with the Classic Editor.

aduth commented on packages/blocks/src/store/reducer.js:
```diff
@@ -20,12 +20,21 @@ import { combineReducers } from '@wordpress/data';
 
 /**
- * Module Constants
+ * @typedef {Object} WPBlockCategory
```
It should probably be descriptive of the _block type_:

```suggestion
 * @typedef {Object} WPBlockTypeCategory
```

aduth replied:
> It should probably be descriptive of the _block type_:

Not going to change. See https://github.com/WordPress/gutenberg/pull/22192#discussion_r421788753.

aduth commented:
Starting to explore optional block category at #22192. It's a bit more of a technical challenge than I'd anticipated it to be.

aduth committed:
Blocks: Normalize block category during registration

aduth committed:
Block Library: Reorganize blocks to new categories

aduth commented:
Rebased again after #22280 was merged, which adds support for category to be omitted from block registration settings (accounting for **future** cases where plugins upgrade to the new categories), if categories are ever updated again.

Updates also include:

- Reorganizing categories per https://github.com/WordPress/gutenberg/pull/19279#issuecomment-624918983, https://github.com/WordPress/gutenberg/pull/19279#issuecomment-624970400, https://github.com/WordPress/gutenberg/pull/19279#issuecomment-625248280
- Removing `getCategory` selector from initial implementation, preferring normalization during block registration (see https://github.com/WordPress/gutenberg/pull/19279#discussion_r424680522)

The end-to-end tests may pass now. Previously, I was seeing issues because the "Slash" block inserter does not search in the same way that the primary block inserter does, and the reorganization / renaming of categories seemed to have an effect on the order of blocks (in a way that the tests were not expecting). Manual testing shows the "Embed" block result appears first now again when searching `/embed`, which may be enough to make the tests happy. There may still be some future work to consider here to make this behavior more predicable and more consistent with the main block inserter search.

aduth committed:
E2E Tests: Update block directory test to use updated category

mtias commented:
Let's get this in, we can reorder a few of the blocks afterwards once we see it in practice for a few days.

aduth commented:
Rebased once again.

ZebulanStanphill reviewed (approved):
Yeah, let's get this in and iterate on where to put certain blocks in future PRs.

aduth committed:
Compat: Reconcile against core-updated categories

Anticipate that core will update to use new categories. Avoid duplicate categories, since legacy categories will not exist to be replaced.

aduth committed:
Compat: Add Trac ticket for block category updates

aduth commented:
I've updated the original comment to reflect the latest state of the pull request, to help those who land here after the pull request is merged.

I've also added the "Needs Dev Note" label, since I think this could be a potential point of confusion for block implementers upgrading to WordPress 5.5.

I created a Trac ticket at https://core.trac.wordpress.org/ticket/50278 , including patch for the necessary changes.

I rebased the branch since there were unrelated failures in the latest version of the `master` branch at the time of the last rebase.

I also pushed some small revisions to the category filtering in 651ff94, since the previous implementation did not properly anticipate that the default categories would be updated in core in the future.

aduth commented:
Proposed devnote text:

>The default set of block categories was updated.
>
>The new set of block categories are:
>
>- Text
>- Media
>- Design
>- Widgets
>- Embeds
>- Reusable
>
>Of the above, a few have remained unchanged: Widgets, Embeds, Reusable.
>
>Mappings have been added to ensure that existing blocks will be assigned to as close as possible to an equivalent category in the new set:
>
>- Common → Text
>- Formatting → Text
>- Layout → Design
>
>Note that if you update your blocks to one of the new block categories, your block may no longer register correctly in older versions of WordPress. If your plugin supports older versions, you should consider to check if a category is registered before using it:
>
>In PHP:
>
>```php
>global $post;
>$has_formatting_category = in_array( 'common', wp_list_pluck( get_block_categories( $post ), 'slug' ) );
>
>register_block_type( 'my-plugin/my-block', array(
>    'category' => $has_formatting_category ? 'formatting' : 'media',
>    
>    // ...
>) );
>```
>
>In JavaScript:
>
>```js
>var hasFormattingCategory = wp.blocks.getCategories().some( function( category ) {
>    return category.slug === 'formatting';
>} );
>
>wp.blocks.registerBlockType( 'my-plugin/my-block', {
>    category: hasFormattingCategory ? 'formatting' : 'media',
>
>    // ...
>} );
>```

gziolo commented:
Nice work. I have a related question. I see that `@wordpress/create-block` was updated already. I'm wondering how it should be tackled in the CHANGELOG there to make it clear that those categories will work only with WordPress 5.5 or with Gutenberg plugin enabled. An alternative is to temporarily add this mapping on the PHP side that is listed in the Dev Note.

shinyabw commented:
Hi everyone. I really appreciate seeing my proposal became a reality. I'm speechless. Thank you very much.

@mtias , for the record, I would like to note that I really liked your proposal using "create, build, share." I did spend time with your idea. I contemplated with "draw, build, share," and "ink, map, let," however, sadly I have gotten busy with COVID-19 situation (I am currently a university prof and admin). 

I feel that your proposal deserves revisiting in the future.

Once again, everyone, thank you very much. I hope I get a chance to meet you someday. If you ever come to Tokyo, please let me know.

With warm regards from Tokyo.

mtias commented:
Oh, thank _you_ @shinyabw for contributing and making things better. I agree there's something to the verb-groupings that'd be worth exploring in the future. Hope you are doing well over there. Have never been to Tokyo but would love to visit some day. Stay safe.
